pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -148,8 +152,32 @@ public synchronized void registerSendOp(CompletableFuture<MessageId> sendFuture)
         }
     }
 
+    public synchronized <T> void registerInternal(CompletableFuture<T> completableFuture) {
+        //There have been an exception, it means this transaction should not be commit,
+        //so there is no need to record the execution of other operations
+        if (!executedFuture.isCompletedExceptionally()) {
+            opsExecutingInTxn.incrementAndGet();
+            executedFuture = new CompletableFuture<>();
```

Comment:
Can achieve the purpose by only using one instance? Looks like,

- A new future was created when the transaction opened.
- when all the ongoing requests are done, the opsExecutingInTxn change to 0, and then compete the future

Reply:
* We cannot confirm that all operations have been completed, when opsExecutingInTxn change to 0.
* It can only means the ops send before are completely, and maybe some ops will come, then the opsExecutingInTxn to n from 0.